### ISSUE TITLE
Dagster Databricks: databricks-sdk version incompatibility

### ISSUE THREAD

**Issue by geoHeil** [dagster-io/dagster#18428]
### Dagster version

0.15.10

### What's the issue?

Based on https://github.com/dagster-io/dagster/discussions/18422 

I conclude that dagsters step launcher has an issue with a cluster which is already existing. And any package might be installed only after the task has already failed.

It would be great if the (missing) packages are installed before launching the job in case of a pre-existing cluster

### Deployment type

Local

### Message from the maintainers

Impacted by this issue? Give it a 👍! We factor engagement into prioritization.

**Comment by geoHeil** [comment]
I think upgrading should be quite painless on dagster side as far as I see. But it looks like for whatever reason databricks themselves is still delivering the old version.

**Comment by smackesey** [comment]
Not a Databricks expert so forgive me for missing something obvious, but I'm a little confused here. Why do `databricks-sdk` in the Dagster environment and Databricks environment need to match?

The Dagster environment uses `databricks-sdk` to build API calls to Databricks. But to my knowledge, the code that you run on Databricks uses `databricks-sdk` (if at all) for a completely different purpose.

**Comment by zyd14** [comment]
Hmm yeah @geoHeil is it possible for you to share the full stack trace of the error you encounter when the versions don't match? I didn't think about it but @smackesey brings up a good point, I'm curious now as to where the ImportError is coming from - is it just from importing `dagster-databricks`? I wonder if the pickled step context object has to import the step launcher module when it gets run, and it's failing when that module tries to import the JobsHealthRule that seems to have moved in newer versions of `databricks-sdk`

**Comment by geoHeil** [comment]
I think it is dagster pyspark what relies on a (more modern version) to be installed

**Comment by zyd14** [comment]
> I think it is dagster pyspark what relies on a (more modern version) to be installed

@geoHeil yes, dagster-pyspark requires <0.0.9, but the question at hand is why is the version mismatch causing an error. The code that is being executed by the step launcher shouldn't directly rely on the databricks-sdk, so the question is where is the error coming from? A stack trace would be really helpful in answering this.

**Comment by geoHeil** [comment]
https://dagster.slack.com/files/U067T8MRS22/F067V5D419U/error.txt

```
AttributeError: module 'databricks.sdk.service.jobs' has no attribute 'JobsHealthRule'
---------------------------------------------------------------------------
AttributeError                            Traceback (most recent call last)
File ~/.ipykernel/2568/command--1-3293539468:13
     10 del sys
     12 with open(filename, "rb") as f:
---> 13   exec(compile(f.read(), filename, 'exec'))

File /tmp/tmpvj4ft08l.py:158
    154                     event_writing_thread.join()
    157 if __name__ == "__main__":
--> 158     main(sys.argv[1], sys.argv[2], sys.argv[3])

File /tmp/tmpvj4ft08l.py:140, in main(step_run_ref_filepath, setup_filepath, dagster_job_zip)
    137 except Exception as e:
    138     # ensure that exceptiosn make their way into stdout
    139     traceback.print_exc()
--> 140     raise e
    141 finally:
    142     # write final stdout and stderr
    143     with open(stderr_filepath, "wb") as handle:

File /tmp/tmpvj4ft08l.py:93, in main(step_run_ref_filepath, setup_filepath, dagster_job_zip)
     91 # We can use regular local filesystem APIs to access DBFS inside the Databricks runtime.
     92 with open(setup_filepath, "rb") as handle:
---> 93     databricks_config = pickle.load(handle)
     95 # sc and dbutils are globally defined in the Databricks runtime.
     96 databricks_config.setup(dbutils, sc)  # noqa: F821  # type: ignore

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/dagster_databricks/__init__.py:13
      1 """Databricks integration for Dagster.
      2 
      3 This package provides:
   (...)
      8     to execute an arbitrary task in Databricks.
      9 """
     11 from dagster._core.libraries import DagsterLibraryRegistry
---> 13 from .databricks import (
     14     DatabricksClient as DatabricksClient,
     15     DatabricksError as DatabricksError,
     16     DatabricksJobRunner as DatabricksJobRunner,
     17 )
     18 from .databricks_pyspark_step_launcher import (
     19     DatabricksConfig as DatabricksConfig,
     20     DatabricksPySparkStepLauncher as DatabricksPySparkStepLauncher,
     21     databricks_pyspark_step_launcher as databricks_pyspark_step_launcher,
     22 )
     23 from .ops import (
     24     create_databricks_run_now_op as create_databricks_run_now_op,
     25     create_databricks_submit_run_op as create_databricks_submit_run_op,
     26 )

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/dagster_databricks/databricks.py:297
    292                 return
    294             time.sleep(poll_interval_sec)
--> 297 class DatabricksJobRunner:
    298     """Submits jobs created using Dagster config to Databricks, and monitors their progress.
    299 
    300     Attributes:
   (...)
    304         max_wait_time_sec (int): How long to wait for a run to complete before failing.
    305     """
    307     def __init__(
    308         self,
    309         host: str,
   (...)
    314         max_wait_time_sec: int = DEFAULT_RUN_MAX_WAIT_TIME_SEC,
    315     ):

File /local_disk0/.ephemeral_nfs/cluster_libraries/python/lib/python3.10/site-packages/dagster_databricks/databricks.py:450, in DatabricksJobRunner()
    426     job_health_settings = self._get_job_health_settings(run_config)
    428     return self.client.workspace_client.jobs.submit(
    429         run_name=run_config.get("run_name"),
    430         tasks=[
   (...)
    445         **notification_settings,
    446     ).bind()["run_id"]
    448 def _get_job_health_settings(
    449     self, run_config: Mapping[str, Any]
--> 450 ) -> Optional[List[jobs.JobsHealthRule]]:
    451     if "job_health_settings" in run_config:
    452         job_health_settings = [
    453             jobs.JobsHealthRule.from_dict(h) for h in run_config["job_health_settings"]
    454         ]

AttributeError: module 'databricks.sdk.service.jobs' has no attribute 'JobsHealthRule'
```

**Comment by zyd14** [comment]
@geoHeil I got curious and tested out databricks-sdk versions 0.8.0-0.17.0 and haven't been able to reproduce this error - do you happen to know what version of the databricks-sdk you were running into this issue with? I'm hoping to be able to make a PR to allow a wider version range - from my testing this morning it seems like we could allow anything <=0.17.0, but I'd like to do some extra testing on whatever version you saw this error on to see if I can reproduce it.

**Comment by geoHeil** [comment]
my testing shows <= 0.17 is fine but >= 0.8 otherwise it fails

**Comment by ion-elgreco** [comment]
@smackesey @danielgafni I think this can be closed

**Comment by danielgafni** [comment]
Step launchers [were superseded by Dagster Pipes](https://github.com/dagster-io/dagster/discussions/25685) and are not under development anymore.
